### PR TITLE
Mixed Bag of Quality and Testing Improvements

### PR DESCRIPTION
**Pull request by irq0**
Mixed bag of quality, testing and quality of life improvements.

Highlights:
 - -Werror, -Wextra+friends
 - const most of types.h
 - Fix struct initialization and shadowing errors

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

### CODE REVIEW

**Comment by irq0** [comment]
The compiler warning commit needs some work. I tested it with clang, but gcc wants other settings / had different defaults:

GCC needs:
-Werror=missing-field-initializers
GCC does not like:
-Wno-error=unqualified-std-cast-call

File: src/rgw/driver/sfs/bucket.cc [hunk 1]
```diff
@@ -47,6 +50,18 @@ void SFSBucket::write_meta(const DoutPrefixProvider* dpp) {
...
+void SFSBucket::Meta::dump(ceph::Formatter* f) const {
+  f->open_object_section("info");
+  info.dump(f);
+  f->close_section();  // info
+                       // encode_json("multipart", multipart, f);
+}
+
+void SFSBucket::Meta::decode_json(JSONObj* obj) {
+  JSONDecoder::decode_json("info", info, obj);
+  // JSONDecoder::decode_json("multipart", multipart, obj);
```
Comment (jecluis) [inline src/rgw/driver/sfs/bucket.cc]: drop the comment

File: src/rgw/driver/sfs/bucket.cc [hunk 2]
```diff
@@ -47,6 +50,18 @@ void SFSBucket::write_meta(const DoutPrefixProvider* dpp) {
   // TODO
 }
 
+void SFSBucket::Meta::dump(ceph::Formatter* f) const {
+  f->open_object_section("info");
+  info.dump(f);
+  f->close_section();  // info
+                       // encode_json("multipart", multipart, f);
```
Comment (jecluis) [inline src/rgw/driver/sfs/bucket.cc]: drop this comment

File: src/rgw/driver/sfs/sfs_lc.h [hunk 3]
```diff
@@ -27,13 +27,13 @@ class LCSFSSerializer : public StoreLCSerializer {
...
-      std::timed_mutex& m, SFStore* store, const std::string& oid,
-      const std::string& lock_name, const std::string& cookie
+      std::timed_mutex& m, SFStore* /* store */, const std::string& /* oid */,
+      const std::string& /* lock_name */, const std::string& /* cookie */
   )
       : mutex(m) {}
 
   virtual int try_lock(
-      const DoutPrefixProvider* dpp, utime_t dur, optional_yield y
+      const DoutPrefixProvider* /* dpp */, utime_t dur, optional_yield /* y */
```
Comment (jecluis) [inline src/rgw/driver/sfs/sfs_lc.h]: drop commented out parameters?

File: src/rgw/driver/sfs/writer.cc [hunk 4]
```diff
@@ -281,11 +282,11 @@ int SFSAtomicWriter::process(bufferlist&& data, uint64_t offset) {
...
 
 int SFSAtomicWriter::complete(
-    size_t accounted_size, const std::string& etag, ceph::real_time* mtime,
+    size_t accounted_size, const std::string& etag, ceph::real_time* out_mtime,
     ceph::real_time set_mtime, std::map<std::string, bufferlist>& attrs,
     ceph::real_time delete_at, const char* if_match, const char* if_nomatch,
-    const std::string* user_data, rgw_zone_set* zones_trace, bool* canceled,
-    optional_yield y
+    const std::string* user_data, rgw_zone_set* /* zones_trace */,
+    bool* canceled, optional_yield /* y */
```
Comment (jecluis) [inline src/rgw/driver/sfs/writer.cc]: drop commented parameter names?

**Review by jecluis — CHANGES_REQUESTED** [review]
few things caught my eye, but generally looks good

**Review by jecluis — APPROVED** [review]

**Comment by irq0** [comment]
Last test failures were s3 test test_lifecycle_expiration_versioned_tags2
Could reproduce the fail on aquarist-labs/s3gw as well for a while. Weird, likely flaky. Issue: https://github.com/aquarist-labs/s3gw/issues/670

**Comment by irq0** [comment]
Reproduced the test fail on aquarist-labs/s3gw with and without this PR. Merging this, to get it out of the way.
Fix for the LC S3 test coming up in https://github.com/aquarist-labs/ceph/pull/186

**Pull request merged**